aten/src/ATen/core/op_registration/op_registration.cpp

```diff
@@ -165,4 +165,42 @@ RegisterOperators::~RegisterOperators() = default;
 RegisterOperators::RegisterOperators(RegisterOperators&&) noexcept = default;
 RegisterOperators& RegisterOperators::operator=(RegisterOperators&&) noexcept = default;
 
+
+CppFunction::CppFunction(KernelFunction func, std::unique_ptr<c10::FunctionSchema> schema)
+  : func_(std::move(func))
+  , schema_(std::move(schema))
+  {}
+
+Module::Module(const char* ns)
+  : ns_(ns)
+  {}
+
+// NB: lives here to reduce link time
+Module::Module(Module&&) = default;
+Module& Module::operator=(Module&&) = default;
+
+
+Module&& Module::def(const char* schema) && {
+  register_.op(c10::RegisterOperators::options().schema(schema));
+  return std::move(*this);
+}
+
+Module&& Module::def(const char* unqual_name, CppFunction&& f) && {
+  // TODO: slow!  Fix internal data structures so I don't have to paste the
+  // names together
+  std::string name;
+  if (ns_) {
+    std::ostringstream oss;
+    oss << ns_ << "::" << unqual_name;
+    name = oss.str();
+  } else {
+    name = unqual_name;
+  }
+  // To be destructively moved out of below
+  register_.op(c10::RegisterOperators::options()
```

Comment:
I'm a bit confused here. How do you handle `.kernel()` vs `.catchAllKernel()` vs `.impl_unboxedOnlyKernel()` vs `.impl_unboxedOnlyCatchAllKernel()`? I think there's some logic missing here.

Reply:
oh nvm, you're calling the private `.kernel()` API here, that means CppFunction should handle this distinction, not the code here. I still think it's missing though.

Reply:
The answer is that it is handled by `TORCH_OPTIMIZED_FN`. So there are two constructors in CppFunction:

```
    // This overload accepts function pointers, e.g., CppFunction(&add_impl)
    template <typename Func>
    explicit CppFunction(Func* f, std::enable_if_t<guts::is_function_type<Func>::value, std::nullptr_t> = nullptr)
      : func_(c10::KernelFunction::makeFromUnboxedRuntimeFunction(f))
      // TODO: Don't go through WrapRuntimeKernelFunctor
      , schema_(detail::FunctionSchemaInferer<detail::WrapRuntimeKernelFunctor<std::decay_t<Func>>>()())
      {}  

```

That's `kernel` and `catchAllKernel`

Then there's

```
    template <typename Func, Func* f>
    static CppFunction makeOptimized() {
      return CppFunction(
        // TODO: For some reason, the "optimized" codepath is still using the
        // RuntimeFunction functor, which is kind of weird
        c10::KernelFunction::makeFromUnboxedOnlyRuntimeFunction(f),
        // NB: disable function schema inference because some ops don't support it yet (TODO fix)
        /* schema */ nullptr
      );
    }
```

That's unboxedOnly

Reply:
Some of the confusion is because I got the names wrong. I'm going to rename `TORCH_OPTIMIZED_FN` to just `CppFunction::makeUnboxedOnly` for clarity.